scipy/optimize/zeros.py

```diff
@@ -173,18 +173,13 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
                 warnings.warn(msg, RuntimeWarning)
                 return p0
             fval = func(*myargs)
-            if fprime2 is not None:
-                fder2 = fprime2(*myargs)
-            if fder2 == 0:
+            if fprime2 is None:
                 # Newton step
                 p = p0 - fval / fder
             else:
-                # Parabolic Halley's method
-                discr = fder ** 2 - 2 * fval * fder2
-                if discr < 0:
-                    p = p0 - fder / fder2
-                else:
-                    p = p0 - 2*fval / (fder + sign(fder) * sqrt(discr))
+                fder2 = fprime2(*myargs)
+                # Halley's method
+                p = p0 - 2 * fval * fder / (2 * fder ** 2 - fval * fder2)
```

Comment:
Probably better to use the update

```
p = p0 - (fval / fder) / (1 - (fvar / fder)*(0.5*fder2 / fder)
```

to avoid potential overflow in the square.

Reply:
I would have agreed, but I tried that first [in this commit] (https://github.com/scipy/scipy/blob/f24a5594351f302f57a82fd813af9b67ea457a6d/scipy/optimize/zeros.py#L183-L184) and _this_ was the form that had overflow issues. I'll try digging into to it more to figure out why. Thanks for looking this over and getting the Wolfram reference on the irrational Halley's form. Very interesting.